### PR TITLE
Update ShimLayer: sync with Roslyn's Syntax.xml

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Sync Syntax.xml to be equal to [Roslyn's Syntax.xml](https://github.com/dotnet/roslyn/blob/Visual-Studio-2022-Version-17.8-Preview-1/src/Compilers/CSharp/Portable/Syntax/Syntax.xml) from Visual-Studio-2022-Version-17.8-Preview-1.
Related to #8184

### CODE REVIEW

File: analyzers/its/expected/akka.net/Akka.Cluster.Sharding--netstandard2.0-S2583.json [hunk 1]
```diff
@@ -1,52 +0,0 @@
-{
-"issues":  [
-{
-"id":  "S2583",
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/its/expected/akka.net/Akka.Cluster.Sharding--netstandard2.0-S2583.json:4]: The issues on this file seem not related to this PR (maybe a FP fix?)

Comment (cristian-ambrosini-sonarsource) [reply analyzers/its/expected/akka.net/Akka.Cluster.Sharding--netstandard2.0-S2583.json]: Rebasing solved the issue.

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Tunnel Vision Laboratories, LLC. All Rights Reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace StyleCop.Analyzers.Lightup
+{
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    public static class TypeDeclarationExtensions
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs:7]:
```suggestion

    public static class TypeDeclarationExtensions
```

File: analyzers/src/SonarAnalyzer.ShimLayer.CodeGeneration/SyntaxLightupGenerator.cs [hunk 3]
```diff
@@ -1237,8 +1237,8 @@ private sealed class NodeData
                 nameof(AnonymousFunctionExpressionSyntax),
                 nameof(LambdaExpressionSyntax),
                 nameof(ParenthesizedLambdaExpressionSyntax),
-                nameof(SimpleLambdaExpressionSyntax)
-                );
+                nameof(SimpleLambdaExpressionSyntax),
+                nameof(ClassDeclarationSyntax));
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.ShimLayer.CodeGeneration/SyntaxLightupGenerator.cs:1241]: Should we include "TypeDeclaration" and "StructDeclaration" as well?

File: analyzers/its/expected/Net5/Net5--net5.0-S2094.json [hunk 4]
```diff
@@ -30,6 +30,19 @@
 "id":  "S2094",
 "message":  "Remove this empty record, write its code or make it an "interface".",
 "location":  {
+"uri":  "https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/its/sources/Net5/Net5/Records.cs#L33",
```
Comment (martin-strecker-sonarsource) [inline analyzers/its/expected/Net5/Net5--net5.0-S2094.json:33]: Why do these show up? These are FPs.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.ShimLayer.CodeGeneration/SyntaxLightupGenerator.cs (hunk 3 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.ShimLayer.CodeGeneration/SyntaxLightupGenerator.cs:1241]: I would not do it beforehand, once it's needed we might extend this list.
Why? From now onward we will create `TypeDeclarationSyntaxWrapper` that will show up together with `TypeDeclarationSyntax` in intellisense hints and available types. This might cause some confusion and why and when to use the wrapper IMO.

**Review by martin-strecker-sonarsource — APPROVED** [review]

**Review by antonioaversa — APPROVED** [review]
LGTM. I have left a minor suggestion and a couple of questions to confirm my understanding of how this fix works.

File: analyzers/src/SonarAnalyzer.ShimLayer.CodeGeneration/SyntaxLightupGenerator.cs [hunk 5]
```diff
@@ -1209,7 +1209,7 @@ public SyntaxData(in GeneratorExecutionContext context, XDocument document)
             {
                 for (var current = node; current is not null; current = this.TryGetNode(current.BaseName))
                 {
-                    if (current.WrapperName is null)
+                    if (current.ExistingType is not null)
                     {
                         // This is not a wrapper
```
[the remark this answers is not in the document]
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.ShimLayer.CodeGeneration/SyntaxLightupGenerator.cs:1214]: Yes it is

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Tunnel Vision Laboratories, LLC. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs:1]: That's a good point, I'm not sure what should be the right copyright either. @martin-strecker-sonarsource what do you think?

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace StyleCop.Analyzers.Lightup
```
[the remark this answers is not in the document]
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs:4]: We used this to be consistent with the other extensions inside the ShimLayer

File: analyzers/src/SonarAnalyzer.ShimLayer.CodeGeneration/SyntaxLightupGenerator.cs [hunk 8]
```diff
@@ -1209,7 +1209,7 @@ public SyntaxData(in GeneratorExecutionContext context, XDocument document)
             {
                 for (var current = node; current is not null; current = this.TryGetNode(current.BaseName))
                 {
-                    if (current.WrapperName is null)
+                    if (current.ExistingType is not null)
```
[the remark this answers is not in the document]
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.ShimLayer.CodeGeneration/SyntaxLightupGenerator.cs:1212]: WrapperName was not working anymore. It was true for `TypeDeclarationSyntax` when we forced it to be included. Therefore the `RecordDeclarationWrapper` base class was set to `BaseTypeDeclaration` instead of `TypeDeclaration`

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs:1]: I have no idea. In the past, I added this header to all the shimlayer files I added, but that doesn't mean it is the right thing to do.

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        private static readonly Func<TypeDeclarationSyntax, ParameterListSyntax> RecordDeclarationParameterListAccessor;
+
+        static TypeDeclarationExtensions()
+        {
+            // TypeDeclarationSyntax.ParameterList was introduced in Roslyn 4.6 (C#12)
+            ParameterListAccessor = LightupHelpers.CreateSyntaxPropertyAccessor<TypeDeclarationSyntax, ParameterListSyntax>(typeof(TypeDeclarationSyntax), nameof(ParameterList));
+            // In earlier versions, the ParameterList was only available on the derived RecordDeclarationSyntax (starting from version 3.7)
+            // To work with version 3.7 to version 4.6 we need to special case the record declaration and access
+            // the parameter list from the derived RecordDeclarationSyntax.
+            if (SyntaxWrapperHelper.GetWrappedType(typeof(RecordDeclarationSyntaxWrapper)) is { } recordDeclaration)
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs:20]: You are correct and the if is wrong. We need to remove the `if`. For versions <3.7 we also need to create the Accessor, but pass `null` for the type. That will enable a "Fallback" accessor which returns null instead of raising.

[the remark this answers is not in the document]
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeDeclarationExtensions.cs:20]: Thanks @martin-strecker-sonarsource!
